griptape/memory/tool/tool_memory.py

```diff
@@ -0,0 +1,140 @@
+    from griptape.tasks import ActionSubtask
+
+
+@define
+class ToolMemory(ToolMemoryActivitiesMixin, ActivityMixin):
+    name: str = field(
+        default=Factory(lambda self: self.__class__.__name__, takes_self=True),
+        kw_only=True,
+    )
+    artifact_storage: dict[Type, BaseArtifactStorage] = field(factory=dict, kw_only=True)
+    namespace_storage: dict[str, BaseArtifactStorage] = field(factory=dict, kw_only=True)
+    namespace_metadata: dict[str, any] = field(factory=dict, kw_only=True)
+
+    @artifact_storage.validator
+    def validate_artifact_storage(self, _, artifact_storage: dict[Type, BaseArtifactStorage]) -> None:
+        seen_types = []
+
+        for storage in artifact_storage.values():
+            if type(storage) in seen_types:
+                raise ValueError("Can't have more than memory storage of the same type")
+
+            seen_types.append(type(storage))
+
+    def get_storage_for(self, artifact: BaseArtifact) -> Optional[BaseArtifactStorage]:
+        find_storage = lambda a: next(
+            (v for k, v in self.artifact_storage.items() if isinstance(a, k)),
+            None
+        )
+
+        if isinstance(artifact, ListArtifact):
+            if artifact.has_items():
+                return find_storage(artifact.value[0])
+            else:
+                return None
+        else:
+            return find_storage(artifact)
+
+    def process_output(
+            self,
+            tool_activity: callable,
+            subtask: ActionSubtask,
+            output_artifact: BaseArtifact
+    ) -> BaseArtifact:
+        from griptape.utils import J2
+
+        tool_name = tool_activity.__self__.name
+        activity_name = tool_activity.name
+        namespace = output_artifact.name
+
+        if output_artifact.is_empty():
+            return InfoArtifact("Tool output is empty.")
+        elif self.store_artifact(namespace, output_artifact):
+            self.namespace_metadata[namespace] = subtask.action_to_json()
+
+            output = J2("memory/tool.j2").render(
+                memory_name=self.name,
+                tool_name=tool_name,
+                activity_name=activity_name,
+                artifact_namespace=namespace
+            )
+
+            return InfoArtifact(output)
+        else:
+            logging.info(f"Output of {tool_name}.{activity_name} can't be stored in {self.name}")
+
+            return ErrorArtifact("Error processing tool output.")
+
+    def store_artifact(self, namespace: str, artifact: BaseArtifact) -> bool:
+        namespace_storage = self.namespace_storage.get(namespace)
+        storage = self.get_storage_for(artifact)
+
+        if namespace_storage and namespace_storage != storage:
+            logging.warning(f"Incompatible storage types")
+
+            return False
+        else:
+            if storage:
+                if isinstance(artifact, ListArtifact):
+                    [
```

Comment:
I know we do this in other areas of the codebase, but I've got mixed feelings on list comprehension without assignment. I think I'm leaning towards encouraging regular for loops.

Reply:
I agree.